### ISSUE TITLE
[BUG] RegexMatchError ?

### ISSUE THREAD

**Issue by Rizmi** [pytube/pytube#1199]
Until now it's worked perfectly, but now it's showing a issue
is this only for me or ?, can anyone help ty

```
Ignoring exception in on_message
Traceback (most recent call last):
File "/.local/lib/python3.9/site-packages/pytube/__main__.py", line 177, in fmt_streams
extract.apply_signature(stream_manifest, self.vid_info, self.js)
File "/.local/lib/python3.9/site-packages/pytube/extract.py", line 409, in apply_signature
cipher = Cipher(js=js)
File "/.local/lib/python3.9/site-packages/pytube/cipher.py", line 33, in __init__
raise RegexMatchError(
pytube.exceptions.RegexMatchError: __init__: could not find match for ^\\w+\\W
```

**Comment by AlexanderMelnikov** [comment]
Same.
Simple code:
```
from pytube import YouTube  
YouTube('https://www.youtube.com/watch?v=dQw4w9WgXcQ').streams.first().download()
```
Raise error:
```
pytube.exceptions.RegexMatchError: __init__: could not find match for ^\w+\W
```

**Comment by mishailovic** [comment]
lmao, got the same error and wanted to open an issue. Literally yesterday everything was fine

**Comment by TheHamkerCat** [comment]
Prolly some new changes from youtube side

**Comment by compsage** [comment]
ditto

**Comment by Lolcker2** [comment]
got the same error, is it because of the new version?
(the error: pytube.exceptions.RegexMatchError: __init__: could not find match for ^\w+\W)

**Comment by tylerweston** [comment]
As close as I can tell this is being thrown from cipher.py which attempts to thwart YouTubes preventative measures for allowing automated video downloading. If YT wised up and are using a tougher challenge method now this could be a tough nut to crack. Lines 30, 31 in cipher.py
```
        var_regex = re.compile(r"^\w+\W")
        var_match = var_regex.search(self.transform_plan[0])
        if not var_match:
            raise RegexMatchError(
                caller="__init__", pattern=var_regex.pattern
            )
```

Looking through the js that the regex is searching is some giant obfuscated mess that I am in no way qualified to deal with so this is where the extent of my help with this issue ends.

**Comment by JavDomGom** [comment]
Same. Raise exception from here: https://github.com/pytube/pytube/blob/master/pytube/cipher.py#L31

![snapshot_issue_1199](https://user-images.githubusercontent.com/41239389/150191053-c48c0df0-f0e2-4db9-aff4-a19b72bdc5d4.png)

The code for ```js``` changed. Reviewing the regular expression is a good start point.

![snapshot_issue_1199_2](https://user-images.githubusercontent.com/41239389/150192739-8176cdda-e0ff-4532-ac0b-808bd0f190bb.png)

**Comment by juanchosaravia** [comment]
I just updated the regex like this and it's working:
`^\$*\w+\W`

![image](https://user-images.githubusercontent.com/5606964/150198074-b76930b6-e2af-469b-a206-b82c9661b314.png)

I'm able to download the videos again.

**Comment by tylerweston** [comment]
> I just updated the regex like this and it's working: `^\$*\w+\W`
> 
> ![image](https://user-images.githubusercontent.com/5606964/150198074-b76930b6-e2af-469b-a206-b82c9661b314.png)
> 
> I'm able to download the videos again.

Beaaauty! I tried but forgot the * between the $ and w! Time to go review my regular expressions! Cheers

**Comment by JavDomGom** [comment]
> I just updated the regex like this and it's working: `^\$*\w+\W`
> 
> ![image](https://user-images.githubusercontent.com/5606964/150198074-b76930b6-e2af-469b-a206-b82c9661b314.png)
> 
> I'm able to download the videos again.

Thank you Juancho, I add this fix to PR #1193

**Comment by arturtamborski** [comment]
@JavDomGom please add another change to your PR:

```diff
# pytube/parser.py, line 152
-    func_regex = re.compile(r"function\([^)]+\)")
+    func_regex = re.compile(r"function\([^)]*\)")
```

this fixes another exception occurring during audio fetching:
```
    yt.streams.get_by_itag(140).download(directory, file_name)
  File "/home/artur/.venv/lib/python3.8/site-packages/pytube/__main__.py", line 292, in streams
    return StreamQuery(self.fmt_streams)
  File "/home/artur/.venv/lib/python3.8/site-packages/pytube/__main__.py", line 177, in fmt_streams
    extract.apply_signature(stream_manifest, self.vid_info, self.js)
  File "/home/artur/.venv/lib/python3.8/site-packages/pytube/extract.py", line 409, in apply_signature
    cipher = Cipher(js=js)
  File "/home/artur/.venv/lib/python3.8/site-packages/pytube/cipher.py", line 44, in __init__
    self.throttling_array = get_throttling_function_array(js)
  File "/home/artur/.venv/lib/python3.8/site-packages/pytube/cipher.py", line 323, in get_throttling_function_array
    str_array = throttling_array_split(array_raw)
  File "/home/artur/.venv/lib/python3.8/site-packages/pytube/parser.py", line 158, in throttling_array_split
    match_start, match_end = match.span()
AttributeError: 'NoneType' object has no attribute 'span'

```

**Comment by boardkeystown** [comment]
HUGE you guys rock. Can't wait for the patch

**Comment by ryankish** [comment]
> Same. Raise exception from here: https://github.com/pytube/pytube/blob/master/pytube/cipher.py#L31
> 
> ![snapshot_issue_1199](https://user-images.githubusercontent.com/41239389/150191053-c48c0df0-f0e2-4db9-aff4-a19b72bdc5d4.png)
> 
> The code for `js` changed. Reviewing the regular expression is a good start point.
> 
> ![snapshot_issue_1199_2](https://user-images.githubusercontent.com/41239389/150192739-8176cdda-e0ff-4532-ac0b-808bd0f190bb.png)

What text editor is this @JavDomGom

**Comment by tylerweston** [comment]
> > Same. Raise exception from here: https://github.com/pytube/pytube/blob/master/pytube/cipher.py#L31
> > ![snapshot_issue_1199](https://user-images.githubusercontent.com/41239389/150191053-c48c0df0-f0e2-4db9-aff4-a19b72bdc5d4.png)
> > The code for `js` changed. Reviewing the regular expression is a good start point.
> > ![snapshot_issue_1199_2](https://user-images.githubusercontent.com/41239389/150192739-8176cdda-e0ff-4532-ac0b-808bd0f190bb.png)
> 
> What text editor is this @JavDomGom

top screenshot at least looks like PyCharm

**Comment by JavDomGom** [comment]
> @JavDomGom please add another change to your PR:
> 
> ```diff
> # pytube/parser.py, line 152
> -    func_regex = re.compile(r"function\([^)]+\)")
> +    func_regex = re.compile(r"function\([^)]*\)")
> ```
> 
> this fixes another exception occurring during audio fetching:
> 
> ```
>     yt.streams.get_by_itag(140).download(directory, file_name)
>   File "/home/artur/.venv/lib/python3.8/site-packages/pytube/__main__.py", line 292, in streams
>     return StreamQuery(self.fmt_streams)
>   File "/home/artur/.venv/lib/python3.8/site-packages/pytube/__main__.py", line 177, in fmt_streams
>     extract.apply_signature(stream_manifest, self.vid_info, self.js)
>   File "/home/artur/.venv/lib/python3.8/site-packages/pytube/extract.py", line 409, in apply_signature
>     cipher = Cipher(js=js)
>   File "/home/artur/.venv/lib/python3.8/site-packages/pytube/cipher.py", line 44, in __init__
>     self.throttling_array = get_throttling_function_array(js)
>   File "/home/artur/.venv/lib/python3.8/site-packages/pytube/cipher.py", line 323, in get_throttling_function_array
>     str_array = throttling_array_split(array_raw)
>   File "/home/artur/.venv/lib/python3.8/site-packages/pytube/parser.py", line 158, in throttling_array_split
>     match_start, match_end = match.span()
> AttributeError: 'NoneType' object has no attribute 'span'
> ```

Hi @arturtamborski, this is already solved in the current version:
https://github.com/JavDomGom/pytube/blob/master/pytube/parser.py#L152

**Comment by JavDomGom** [comment]
> HUGE you guys rock. Can't wait for the patch

Yes, there are many PR to review and merge. As soon as the original code is up to date we should all update as well. Temporarily, whoever wants can use my fork with these problems already solved and if you prefer you can also apply the changes to your local version.

![snapshot_issue_1199_3](https://user-images.githubusercontent.com/41239389/150292808-e760517b-0538-4859-bf65-a7026d6c5c92.png)

**Comment by JavDomGom** [comment]
> What text editor is this @JavDomGom

PyCharm 2021.3.1 (Professional Edition), IMHO it is the best Python IDE, I use it like a sir. ;D

**Comment by ror1212** [comment]
Update the library with the changes you made and then later we solve the problem of the kit... We want a solution All telegram bots and Windows programs are stopped .. Update the library with what you have solved

**Comment by rmerzouki** [comment]
> I just updated the regex like this and it's working: `^\$*\w+\W`
> 
> ![image](https://user-images.githubusercontent.com/5606964/150198074-b76930b6-e2af-469b-a206-b82c9661b314.png)
> 
> I'm able to download the videos again.

Sure it works perfectly, I tested it locally, but what if the app is deployed on Streamlit Cloud/sharing, Heroku ... ?

**Comment by JavierDominguezGomez** [comment]
> Sure it works perfectly, I tested it locally, but what if the app is deployed on Streamlit Cloud/sharing, Heroku ... ?

https://github.com/pytube/pytube/issues/1199#issuecomment-1017190777

**Comment by rishabh3354** [comment]
Please Anyone share the branch in which it is patched

**Comment by rishabh3354** [comment]
> > HUGE you guys rock. Can't wait for the patch
> 
> Yes, there are many PR to review and merge. As soon as the original code is up to date we should all update as well. Temporarily, whoever wants can use my fork with these problems already solved and if you prefer you can also apply the changes to your local version.
> 
> ![snapshot_issue_1199_3](https://user-images.githubusercontent.com/41239389/150292808-e760517b-0538-4859-bf65-a7026d6c5c92.png)

great work, thankyou so much!!!

**Comment by YitzhakSp** [comment]
what worked for me was:
in pytube/cipher.py change the line:
`var_regex = re.compile(r"^\w+\W")`
to
`var_regex = re.compile(r"^\$*\w+\W")`

**Comment by rmerzouki** [comment]
> what worked for me was:
> in pytube/cipher.py change the line:
> `var_regex = re.compile(r"^\w+\W")`
> to
> `var_regex = re.compile(r"^\$*\w+\W")`

Same for me

**Comment by ehsanbehdad** [comment]
> 

Hi, anyone who's using this library on some cloud-based platform like heroku can use my temporary repository till the next update.
just in requirements.txt file replace pytube with git+https://github.com/ehsanbehdad/Pytube
then you'll be fine.

**Comment by rmerzouki** [comment]
> Hi, you can use my temporary repository for your heroku app till the next update.
> just in requirements.txt file replace pytube with git+https://github.com/ehsanbehdad/Pytube
> then you'll be fine.

Thank you Ehsan that's great will try that !

**Comment by sharmas1ddharth** [comment]
> > 
> 
> Hi, anyone who's using this library on some cloud-based platform like heroku can use my temporary repository till the next update. just in requirements.txt file replace pytube with git+https://github.com/ehsanbehdad/Pytube then you'll be fine.

Thank you @ehsanbehdad, this worked for me!

**Comment by boardkeystown** [comment]
I'm not familiar with getting installing a python package without pip.... 

Could someone explain to me how 

pip install pytube 

set the cli pytube?

**Comment by Rizmi** [comment]
Thank y'all for all the help

**Comment by cimeriann** [comment]
> As close as I can tell this is being thrown from cipher.py which attempts to thwart YouTubes preventative measures for allowing automated video downloading. If YT wised up and are using a tougher challenge method now this could be a tough nut to crack. Lines 30, 31 in cipher.py
> 
> ```
>         var_regex = re.compile(r"^\w+\W")
>         var_match = var_regex.search(self.transform_plan[0])
>         if not var_match:
>             raise RegexMatchError(
>                 caller="__init__", pattern=var_regex.pattern
>             )
> ```
> 
> Looking through the js that the regex is searching is some giant obfuscated mess that I am in no way qualified to deal with so this is where the extent of my help with this issue ends.

how did you open cipher.py

**Comment by tylerweston** [comment]
> > As close as I can tell this is being thrown from cipher.py which attempts to thwart YouTubes preventative measures for allowing automated video downloading. If YT wised up and are using a tougher challenge method now this could be a tough nut to crack. Lines 30, 31 in cipher.py
> > ```
> >         var_regex = re.compile(r"^\w+\W")
> >         var_match = var_regex.search(self.transform_plan[0])
> >         if not var_match:
> >             raise RegexMatchError(
> >                 caller="__init__", pattern=var_regex.pattern
> >             )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looking through the js that the regex is searching is some giant obfuscated mess that I am in no way qualified to deal with so this is where the extent of my help with this issue ends.
> 
> how did you open cipher.py

I have this project setup using venv so cipher is located in
`project_root\venv\Lib\site-packages\pytube\cipher.py`
Note that I named my venv `venv` when I created it, so it might look different for you. In VS Code you can also right click the word `pytube` where you `import pytube`  choose `Go to Definition...` (or F12) and it should bring you to the `__init__.py` in the root of the pytube library and cipher.py should be in the same folder. I'm sure there is something similar for other virtual environments. 
You can also view the source of cipher.py [here](https://github.com/pytube/pytube/blob/master/pytube/cipher.py) if you just want to read through it.

**Comment by cimeriann** [comment]
> 

Thanks man,

I eventually had to find the path to where installed pytube. Cipher.py was in the same folder

**Comment by tylerweston** [comment]
> > 
> 
> Thanks man,
> 
> I eventually had to find the path to where installed pytube. Cipher.py was in the same folder

Yes, I should've mentioned you can do `pip show pytube` to find where a package is installed, under Location
```
~#@❯  pip show pytube                                                                                                                                                               ❮  
Name: pytube
Version: 11.0.1
Summary: Python 3 library for downloading YouTube Videos.
Home-page: https://github.com/pytube/pytube
Author: Ronnie Ghose, Taylor Fox Dahlin, Nick Ficano
Author-email: hey@pytube.io
License: The Unlicense (Unlicense)
Location: c:\[...omitted...]\lib\site-packages
Requires:
Required-by:
```

**Comment by rm112538** [comment]
![image](https://user-images.githubusercontent.com/64090011/150887655-111c3d8e-aa72-4e15-86ae-eec37256adb4.png)

This malfunction was quite a learning experience for me. I downloaded pytube .tar.gz file from PyPi. Then I placed it on my vscode  workspace.  After that I extracted the whole package to my works space using the following code.

`   import tarfile
    tar = tarfile.open("sample.tar.gz")
    tar.extractall()
    tar.close()
`
At this point I was able to make the changes that were suggested in this thread. However, I still have one thing I haven't figured out.  How do I compile the package again to be used in my environment as if I performed pip install? I tried to run "python setup.py install" but "from pytube import YouTube" still had squiggly lines. How do I move forward?

As you can tell my I am a newbie in the world of Python.

**Comment by tylerweston** [comment]
> How do I move forward?

while some Python modules do have to be compiled to be used, I don't think pytube is one of them.  As far as I know it's usually modules that use other, non-interpreted languages that must be compiled. Look into modules and importing them and it should be fine. The way paths work in modules seems a bit finnicky and I don't claim to understand it well, hence why I suggest going and learning about them yourselves. But ya, as far as I know, just importing the Python should work fine. Maybe just try moving the `pytube` folder to the root of your project folder and start from there?

**Comment by rm112538** [comment]
Thank you for your email. I import modules all the time and use their
attributes.
Thank you so much.

**Comment by rishabh3354** [comment]
is pytube is working for anyone, Please share pytube patched branch so that we will be able to install through pip

**Comment by rmerzouki** [comment]
Same issue for me @rishabh3354, did you manage to fix it ?

**Comment by rishabh3354** [comment]
Nope, i shifted my project to ytdlp ( ytdl forked, more advanced and get frequently updates ) . I got sick of pytube and its bugs on daily basis, my project got affected very bad with pytube. Now i am happy with more reliable and powerful lib ytdlp. 
Again i am not saying pytube is bad lib, But due to lack of updates by author and tons of bugs which is not getting fixed by months , i just shifted also i advise you same whosoever is running their project with pytube.

Thanks , this is not hate message for pytube, its just a my recommendation.

**Comment by enragedsweat** [comment]
Some guy showed this image. Which file is that, and where can i access it 
![image](https://user-images.githubusercontent.com/104448363/171636740-5c20e69d-712d-4660-9757-92433a8dfd8b.png)

**Comment by mishailovic** [comment]
> Some guy showed this image. Which file is that, and where can i access it ![image](https://user-images.githubusercontent.com/104448363/171636740-5c20e69d-712d-4660-9757-92433a8dfd8b.png)

https://github.com/pytube/pytube/blob/2e307d8d615ef30aa837fe2275954146dab07ca6/pytube/cipher.py#L30

**Comment by abanand132** [comment]
Hii,
this issue hasn't solved yet. I downloaded this module in pycharm and this "var_regex = re.compile(r"^\$*\w+\W")" is not there in cipher.py file. I have to manually figure it out.
In the case, If I send my code to others then it will through an error in their system. Suggest a way to figure it out ?

**Comment by brilliant-ember** [comment]
I wrote the code to fix this bug in this PR https://github.com/pytube/pytube/pull/1336 I tested it locally and it worked for me.
Until this PR gets merged you can manually edit the cipher file on line 30 and change the regular exp as mentioned in this stack overflow answer https://stackoverflow.com/a/70777385

**Comment by NathanDai5287** [comment]
After I changed this line, the error changed to `AttributeError: 'NoneType' object has no attribute 'span'`

**Comment by ChrisZhangJin** [comment]
May i have a query that when it will have a formal release or update for this fix?

**Comment by jaybhavsaar** [comment]
Hello Programmers,
I am facing this issue in this Django code. can you help me what's my mistakes.

**Code**
from django.shortcuts import render
from pytube import YouTube

def index(request):
    link = str(request.GET)
    print(link)
    youtube_1 = YouTube(link)
    videos = youtube_1.streams.filter(progressive=True, file_extension='mp4', res="720p")
    videos.download()
    print("Downloading...")
    print("Successfully")
    return render(request, "index.html")

**Error**
_In HTML Page_
regex_search: could not find match for (?:v=|\/)([0-9A-Za-z_-]{11}).*

_In Terminal_
pytube.exceptions.RegexMatchError: regex_search: could not find match for (?:v=|\/)([0-9A-Za-z_-]{11}).*

**Comment by igorsantos314** [comment]
> I just updated the regex like this and it's working: `^\$*\w+\W`
> 
> ![image](https://user-images.githubusercontent.com/5606964/150198074-b76930b6-e2af-469b-a206-b82c9661b314.png)
> 
> I'm able to download the videos again.

This solution work for me

**Comment by medhsv** [comment]
i did change line 30 manually, still not fixed.

pytube version 15

yesterday everything was working fine, but today suddenly this issue popped.

Please provide another fix

**Comment by medhsv** [comment]
> > I just updated the regex like this and it's working: `^\$*\w+\W`
> > ![image](https://user-images.githubusercontent.com/5606964/150198074-b76930b6-e2af-469b-a206-b82c9661b314.png)
> > I'm able to download the videos again.
> 
> This solution work for me

I use Google Colab, made the change in the file.
i re run the cell
not fixed.

if am doing anything wrong please correct me

**Comment by igorsantos314** [comment]
> > > I just updated the regex like this and it's working: `^\$*\w+\W`
> > > ![image](https://user-images.githubusercontent.com/5606964/150198074-b76930b6-e2af-469b-a206-b82c9661b314.png)
> > > I'm able to download the videos again.
> > 
> > 
> > This solution work for me
> 
> I use Google Colab, made the change in the file. i re run the cell not fixed.
> 
> if am doing anything wrong please correct me

While the official solution has not been released, I made the correction on the line where the error occurs. Which is what you did, I hope the solution comes out as soon as possible.

I need to generate a docker image and download the updated dependencies for the generated image, without the official fix we have a problem when running the docker build because with the current version **15.0 the Regex problem** will still persist, so to get around this situation I will have to include the venv from my machine, but this is not a good practice.

**Comment by Scylla2020** [comment]
> > HUGE you guys rock. Can't wait for the patch
> 
> Yes, there are many PR to review and merge. As soon as the original code is up to date we should all update as well. Temporarily, whoever wants can use my fork with these problems already solved and if you prefer you can also apply the changes to your local version.
> 
> ![snapshot_issue_1199_3](https://user-images.githubusercontent.com/41239389/150292808-e760517b-0538-4859-bf65-a7026d6c5c92.png)

Unfortunately the fork now throws an `HTTP Error 400: Bad Request`. I've installed the latest pytube and edited cipher.py to get everything working. Anyone interested can install the latest working version via `pip install git+https://github.com/Scylla2020/pytube.git`
